keyboards/ploopyco/trackball_thumb/trackball_thumb.c

```diff
@@ -0,0 +1,291 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "trackball_thumb.h"
+
+#ifndef OPT_DEBOUNCE
+#    define OPT_DEBOUNCE 5  // (ms) 			Time between scroll events
+#endif
+#ifndef SCROLL_BUTT_DEBOUNCE
+#    define SCROLL_BUTT_DEBOUNCE 100  // (ms) 			Time between scroll events
+#endif
+#ifndef OPT_THRES
+#    define OPT_THRES 150  // (0-1024) 	Threshold for actication
+#endif
+#ifndef OPT_SCALE
+#    define OPT_SCALE 1  // Multiplier for wheel
+#endif
+#ifndef PLOOPY_DPI_OPTIONS
+#    define PLOOPY_DPI_OPTIONS { 1200, 1600, 2400 }
+#    ifndef PLOOPY_DPI_DEFAULT
+#        define PLOOPY_DPI_DEFAULT 1
+#    endif
+#endif
+#ifndef PLOOPY_DPI_DEFAULT
+#    define PLOOPY_DPI_DEFAULT 0
+#endif
+#ifndef PLOOPY_DRAGSCROLL_DPI
+#    define PLOOPY_DRAGSCROLL_DPI 100 // Fixed-DPI Drag Scroll
+#endif
+#ifndef PLOOPY_DRAGSCROLL_MULTIPLIER
+#    define PLOOPY_DRAGSCROLL_MULTIPLIER 0.75 // Variable-DPI Drag Scroll
+#endif
+
+keyboard_config_t keyboard_config;
+uint16_t          dpi_array[] = PLOOPY_DPI_OPTIONS;
+#define DPI_OPTION_SIZE (sizeof(dpi_array) / sizeof(uint16_t))
+
+// TODO: Implement libinput profiles
+// https://wayland.freedesktop.org/libinput/doc/latest/pointer-acceleration.html
+// Compile time accel selection
+// Valid options are ACC_NONE, ACC_LINEAR, ACC_CUSTOM, ACC_QUADRATIC
+
+// Trackball State
+bool     is_scroll_clicked = false;
+bool     BurstState        = false;  // init burst state for Trackball module
+uint16_t MotionStart       = 0;      // Timer for accel, 0 is resting state
+uint16_t lastScroll        = 0;      // Previous confirmed wheel event
+uint16_t lastMidClick      = 0;      // Stops scrollwheel from being read if it was pressed
+uint8_t  OptLowPin         = OPT_ENC1;
+bool     debug_encoder     = false;
+bool     is_drag_scroll    = false;
+
+__attribute__((weak)) void process_wheel_user(report_mouse_t* mouse_report, int16_t h, int16_t v) {
+    mouse_report->h = h;
+    mouse_report->v = v;
+}
+
+__attribute__((weak)) void process_wheel(report_mouse_t* mouse_report) {
+    // TODO: Replace this with interrupt driven code,  polling is S L O W
+    // Lovingly ripped from the Ploopy Source
+
+    // If the mouse wheel was just released, do not scroll.
+    if (timer_elapsed(lastMidClick) < SCROLL_BUTT_DEBOUNCE) {
+        return;
+    }
+
+    // Limit the number of scrolls per unit time.
+    if (timer_elapsed(lastScroll) < OPT_DEBOUNCE) {
+        return;
+    }
+
+    // Don't scroll if the middle button is depressed.
+    if (is_scroll_clicked) {
+#ifndef IGNORE_SCROLL_CLICK
+        return;
+#endif
+    }
+
+    lastScroll  = timer_read();
+    uint16_t p1 = adc_read(OPT_ENC1_MUX);
+    uint16_t p2 = adc_read(OPT_ENC2_MUX);
+    if (debug_encoder) dprintf("OPT1: %d, OPT2: %d\n", p1, p2);
+
+    int dir = opt_encoder_handler(p1, p2);
+
+    if (dir == 0) return;
+    process_wheel_user(mouse_report, mouse_report->h, (int)(mouse_report->v + (dir * OPT_SCALE)));
+}
```

Comment:
This allows us to use the encoder functions instead of hard coding the behavior to mouse movements.  

```suggestion
__attribute__((weak)) bool encoder_update_user(uint8_t index, bool clockwise) { return true; }

bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) {
        return false;
    }
#ifdef MOUSEKEY_ENABLE
    tap_code(clockwise ? KC_WH_U : KC_WH_D);
#else
    mouse_report_t mouse_report = pointing_device_get_report();
    mouse_report.v = clockwise ? 1 : -1;
    pointing_device_set_report(mouse_report);
    pointing_device_send();
#endif
    return true;
}

void process_wheel(void) {
    // Lovingly ripped from the Ploopy Source

    // If the mouse wheel was just released, do not scroll.
    if (timer_elapsed(lastMidClick) < SCROLL_BUTT_DEBOUNCE) {
        return;
    }

    // Limit the number of scrolls per unit time.
    if (timer_elapsed(lastScroll) < OPT_DEBOUNCE) {
        return;
    }

    // Don't scroll if the middle button is depressed.
    if (is_scroll_clicked) {
#ifndef IGNORE_SCROLL_CLICK
        return;
#endif
    }

    lastScroll  = timer_read();
    uint16_t p1 = adc_read(OPT_ENC1_MUX);
    uint16_t p2 = adc_read(OPT_ENC2_MUX);
    if (debug_encoder) dprintf("OPT1: %d, OPT2: %d\n", p1, p2);

    int dir = opt_encoder_handler(p1, p2);

    if (dir == 0) return;
    encoder_update_kb(0, dir == 1);
}
```